man/man3/ml_send.3
.TH ML_SEND 3
'ta .8i 1.6i 2.4i 3.2i 4.0i 4.8i 5.6i 6.3i
.SH NAME
Simple Mail Submission (ml_)
.SH SYNOPSIS

.nf
#include "util.h"
#include <stdio.h>

ml_init (return, sender, from, subject)
    int return,
        sender;
    char *from,
         *subject;
ml_to ();
ml_cc ();
ml_adr (address)
    char *address;
ml_aend ();
ml_tinit ();
ml_txt (text);
    char *text;
ml_file (fil-text);
    FILE *file-text;
ml_end (end-stat)
    int  end-stat;
ml_1adr (return, sender, from, subject, address);
.fi
.RE
.SH DESCRIPTION
.PP
These procedures may be used for simple mail submission, providing
a simple interface to the \fImm_\fR package described in mmdf(3).
The headers created by the \fIml_\fR procedures are compatible with
those produced by the \fIv6mail\fR(1) program.
The basic sequence is:

.RS
.nf
# include "util.h"

ml_init (YES, NO, "My Name", "The Subject");
ml_adr ("destination address 1");
ml_adr ("destination address 2");
...
ml_aend ();
ml_tinit ();
ml_txt ("Some opening text");
ml_txt ("maybe some more text");
ml_file (stdio-file-stream-descriptor-of-file-to-include);
if (ml_end (OK)) != OK)
{   error-handling code }

.fi
Arguments that are to be defaulted should be zero.  All routines
return a status value of OK or NOTOK.
.RE
.PP
ml_init()'s arguments specify:
.IP "a)"
whether return-mail (to the sender) should be allowed (1=yes, 0=no),
.IP "b)"
whether a Sender field should be used to specify the
correct sender (contingent on next argument),
.IP "c)"
text for the From field, and
.IP "d)"
text for the Subject field.
.PP
If (b) is NO, then (c)'s text must contain the
correct sender information, or else just contain the "name" portion
of the address.  
.PP
Ml_to() and ml_cc() are used to switch between "To:" and "cc:"
addresses.  Normally, only "To:" addresses are used and, for this,
no ml_to() call is needed.
.PP
Ml_adr() specifies the text of one address.
An "address" is whatever is valid for your system as if you
were typing it to the \fIcheckaddr\fR command.  Ml_aend() signals the end
of address specification, and ml_tinit() signals the start of
message text submission.
.PP
Ml_txt() directly enters raw text.  Ml_file() passes a stdio
file pointer, to be used for file text submission; all of the
(remaining) contents of the file are submitted.
Any number of ml_txt() and ml_file() calls may be made and they
may be freely mixed. They just
append text to the message.  The text must contain its own
newlines.
.PP
Ml_1adr() is provided as a convenient way to initiate mail
when there is only one addressee.  Its arguments are the same
as for ml_init(), except there is an extra argument for
the address string.
It may be used, with the earlier example,
to replace the ml_init(), ml_addr(), ml_aend(), and ml_tinit()
calls.
.SH FILES
.IP "lib/libmmdf.a" 25
contains the ml_send library.
.SH "SEE ALSO"
\fIsubmit\fR(8), mmdf(3), \fIv6mail\fR(1)


Changes
Changes beta 3
	* Small bug in ap_lex.c hip@a.cs.okstate.edu
	* jromine@ics.uci.edu some changes for Solaris 2.5, Sun Pro
	  4.2 Compiler
	* linux 2.2.7, glibc2.1, egcs 2.91 (redhat 6.0)
	* sco (first attempt)
	* jromine@ics.uci.edu, boundary fix smtpsrvr
	* Small bug in submit/lnk_getaddr.c krueger@mathematik.uni-kl.de

Wed Dec 31 13:14:34 MET 1997	Kai Krueger <krueger@mathematik.uni-kl.de>

	* create file h/config.h.in and moved every define to it.

	* DOASSIGN renamed to HAVE_ASSIGN. Autodetection by configure

	* SECURETTY renamed to HAVE_SECURETTY, Autodetection by configure

	* NODIAL renamed to HAVE_DIAL. default is 'no'
	  may be enanble by configure --enable-dial
	  if compiling phone/smphone channel it wil be turned on by configure.

	* NAMESERVER renamed to HAVE_NAMESERVER. Autodetection by configure.
	  May be overwriten by --disable-nameserver
          - tb_ns.4.2.c and tb_ns.fake.c merged into one file tb_ns.c and
	  seperated by #ifdef HAVE_NAMESERVER...
	  - removed TB_NS in Makefile.com	  

	* HAVE_TCP_WRAPPER renamed to HAVE_LIBWRAP. Autodetection by configure.
	
	* <time.h> included in config.h. It's location is determined by
	  configure

	* determination of <dir.h>, <dirent.h>, a.s.o is done by configure
	  definition an include has moved to config.h

	* determination of index/rindex done by configure and moved to config.h

	* definition of sigtype renamed to RETSIGTYPE. Detection done by
	  configure.

	* return-type of sprintf determined by configure. Define BSD_SPRINTF
	  if char * is returned.

	* uip/Makefile, src/Makefile, lib/Makefile Makefile.com and Makefile
	  are generated by configure.

	* Value of CH_TB in Makefile.com set by configure. If you want
	  sequential tables run configure with the argument --disable-dbm

	* Added define HAVE_SGTTY_H for the existens of <sgtty.h>

	* Added check for <varargs.h> as define HAVE_VARARGS_H.
	  Added check for variable number of arguments if <varargs.h> don't
	  exists. If Compiler don't supports variable number of arguments
	  NO_VARARGS is defined. (NO_VARARGS should be renamed)
	  configure ist checking for both of it. But if <varagrs.h> isn't
	  found lib/util/multcat.c and lib/util/multcpy.c could't be compiled.
	  
	* test configure on machine that dosn't support multiple arguments.
	  (check of NO_VARARGS)

	* test for dup2() and fcntl(fd1, F_DUPFD, fd2) done.
	  NODUP2 replaced by HAVE_DUP2 and
	  NOFCTNL replaced by HAVE_FCNTL_F_DUPFD


lib/mmdf/ml_send.c
#include "util.h"
#include "mmdf.h"
#include "cnvtdate.h"
#include <pwd.h>
#include "ml_send.h"
#include "mm_io.h"

/*
 *     MULTI-CHANNEL MEMO DISTRIBUTION FACILITY  (MMDF)
 *     
 *
 *     Copyright (C) 1979,1980,1981  University of Delaware
 *     
 *     Department of Electrical Engineering
 *     University of Delaware
 *     Newark, Delaware  19711
 *
 *     Phone:  (302) 738-1163
 *     
 *     
 *     This program module was developed as part of the University
 *     of Delaware's Multi-Channel Memo Distribution Facility (MMDF).
 *     
 *     Acquisition, use, and distribution of this module and its listings
 *     are subject restricted to the terms of a license agreement.
 *     Documents describing systems using this module must cite its source.
 *
 *     The above statements must be retained with all copies of this
 *     program and may not be removed without the consent of the
 *     University of Delaware.
 *     
 *
 *     version  -1    David H. Crocker    March   1979
 *     version   0    David H. Crocker    April   1980
 *     version  v7    David H. Crocker    May     1981
 *     version   1    David H. Crocker    October 1981
 *
 */
/* send a piece of mail, using a simple interface to submit */

/*  Basic sequence is:
 *
 *          ml_init (YES, NO, "My Name", "The Subject");
 *          ml_adr ("destination address 1");
 *          ml_adr ("destination address 2");
 *          ...
 *          ml_aend ();
 *          ml_tinit ();
 *          ml_txt ("Some opening text");
 *          ml_txt ("maybe some more text");
 *          ml_file (file-stream-descriptor-of-file-to-include);
 *          if (ml_end (OK)) != OK)
 *          {   error-handling code }
 *
 *  Arguments that are to be defaulted should be zero.
 *
 *  ml_init's arguments specify a) whether return-mail (to the sender
 *  should be allowed, b) whether a Sender field should be used to
 *  specify the correct sender (contingent on next argument), c) text
 *  for the From field, and d) text for the Subject field.  If (b) is
 *  NO, then (c)'s text will be followed by the correct sender
 *  information.
 *
 *  ml_to and ml_cc are used to switch between To and CC addresses.
 *  Normally, only To addresses are used and, for this, no ml_to call is
 *  needed.
 *
 *  An "address" is whatever is valid for your system, as if you were
 *  typing it to the mail command.
 *
 *  You may freely mix ml_txt and ml_file calls.  They just append text
 *  to the message.  The text must contain its own newlines.
 *
 *  State diagram: ml_state:
 *       ML_FRESH --> submit running --> ML_MSG  --> ready for new message -->
 *       ML_HEADER --> reading header --> ML_TEXT --> message sent --> ML_MSG
 *
 */

#include <signal.h>

extern struct ll_struct   *logptr;
extern char *locfullname;
#ifndef __STDC__
extern struct passwd *getpwuid ();
#endif
extern char *getmailid ();

int ml_state = ML_FRESH;

static char *fieldname;
static char *logdir;
static char *username;

/**/

int ml_init (ret, sndr, from, sub)    /* set-up for using submit            */
int     ret,                      /* allow return mail to sender?       */
	sndr;                     /* include Sender field?              */
char    *sub,                     /* subject line                       */
        *from;			  /* from field                         */
{
    static first_ml = TRUE;
    char linebuf[LINESIZE];
    char datbuf[64];
    struct passwd  *pwdptr;
    int     realid,
	    effecid;

    if (first_ml) {
        getwho (&realid, &effecid);   /* who am i?                          */
        
        if ((pwdptr = getpwuid (effecid)) == (struct passwd *) NULL)
            return(NOTOK);
        
        /*  save login directory   */
        logdir = strdup(pwdptr -> pw_dir);
        
        if ((username = getmailid(pwdptr -> pw_name)) == NULL)
            return(NOTOK);

	first_ml = FALSE;
    }

    if (ml_state != ML_FRESH && ml_state != ML_MSG) {  /* odd state; resync */
	mm_end(NOTOK);
	ml_state = ML_FRESH;
    }

    if (ml_state == ML_FRESH) {
	if (rp_isbad(mm_init()) || rp_isbad(mm_sbinit()))
	    return(NOTOK);
	ml_state = ML_MSG;
    }

    strcpy (linebuf, "rmxto,cc*");
    if (!ret)                     /* disable return to sender           */
	strcat(linebuf, "q");

    if (rp_isbad(mm_winit((char *) 0, linebuf, (char *) 0))) {
	ml_end(NOTOK);
	return(NOTOK);
    }
    ml_state = ML_HEADER; 

    cnvtdate (TIMREG, datbuf);
    ml_sndhdr("Date:", datbuf);

    if (from != 0)
    {                             /* user-specified From field          */
	if (sndr) {
	    ml_sndhdr("From:", from);
	    ml_sender("Sender:", (char *) 0, TRUE);
	}
	else
	    ml_sender("From:", from, TRUE);
    }
    else
        ml_sender("From:", (char *) 0, TRUE);

    if (sub != 0) 		  /* user-specified Subject field       */
	ml_sndhdr("Subject:", sub);

    return (ml_to ());		  /* set-up for To: addresses           */
}
/**/


int ml_to ()			  /* ready to specify To: address       */
{
    fieldname = "To:";
    return (OK);
}

int ml_cc ()			  /* ready to specify CC: address       */
{
    fieldname = "Cc:";
    return (OK);
}

int ml_adr (address)		  /* a destination for the mail         */
char    address[];
{
    ml_sndhdr (fieldname, address);
    return (OK);
}

int ml_aend ()			  /* end of addrs                       */
{
    if (ml_state == ML_HEADER) 
	mm_wtxt("\n", 1);
    return (OK);
}

int ml_tinit ()                     /* ready to send text                 */
{
    if (ml_state == ML_HEADER)
	ml_state = ML_TEXT;
    return (OK);
}

int ml_sndhdr (name, contents)
char *name;
char *contents;
{
    char linebuf[LINESIZE];

    if (ml_state == ML_HEADER) {
	sprintf(linebuf, "%-10s%s\n", name, contents);
	mm_wtxt(linebuf, strlen (linebuf));
    }
    return(OK);
}
/**/
void ml_sender (cmpnt, name)
char    cmpnt[],
	name[];
{
    int     sigfd,
	    sigsiz;
    char   *ptr,
	    linebuf[LINESIZE],
	    sigtxt[LINESIZE];     /* where is signature text?           */
    char    gotsig;

    gotsig = FALSE;

    if (name != 0)
    {
	strcpy (sigtxt, name);
	gotsig = TRUE;
    }
    else
    {                             /* user didn't give us a signature    */
	sprintf (linebuf, "%s/.signature", logdir);

	if ((sigfd = open (linebuf, 0)) >= 0)
	{                         /* there is a file w/signature?       */
	    if ((sigsiz = read (sigfd, sigtxt, sizeof sigtxt)) > 0) {
	    	char *cp;

		sigtxt[sigsiz - 1] = '\0';
	    	if ((cp = strchr(sigtxt, '\n')))
	    	    *cp = '\0';
	    	if (sigtxt[0])
		    gotsig = TRUE;
	    }
	}
    }

    if (gotsig)                   /* real name + mailbox                */
    {
	for (ptr = locfullname; *ptr != 0; ptr++)
	    *ptr = uptolow (*ptr);      /* Screw locfullname */
	sprintf (linebuf, "%s <%s@%s>", sigtxt, username, locfullname);
    }
    else                          /* just the mailbox info              */
	sprintf (linebuf, "%s@%s", username, locfullname);

    ml_sndhdr (cmpnt, linebuf);
}

/*  */
int ml_file (infp)                    /* send a file to the message         */
register FILE  *infp;             /* input stdio file stream pointer    */
{
    register short len;
    char    buffer[BUFSIZ];

    if (ml_state != ML_TEXT)
	return (OK);

    while ((len = fread (buffer, sizeof (char), sizeof(buffer), infp )) > 0) {
	if (rp_isbad(mm_wtxt(buffer, len))) {
	    ll_log (logptr, LLOGFAT, "ml_file() mm_wtxt error");
	    ml_end (NOTOK);  
	    return (NOTOK);
	}
    }

    if (!feof(infp))
    {
	ll_log (logptr, LLOGFAT, "ml_file() Error on file");
	ml_end (NOTOK);
	return (NOTOK);
    }
    return (OK);
}

int ml_txt (text)                     /* some text for the body             */
char text[];                      /* the text                           */
{
    if (ml_state != ML_TEXT)
	return (OK);

    if (rp_isbad(mm_wtxt(text, strlen(text))))
    {
	ml_end (NOTOK);
	return (NOTOK);
    }
    return (OK);
}
/**/


int ml_end (type)			  /* message is finished                */
int     type;                     /* normal ending or not               */
{
    struct rp_bufstruct thereply;
    int len;

    switch (type) {
	case OK:
	    if (ml_state != ML_TEXT || rp_isbad(mm_wtend()) 
				 || rp_isbad(mm_rrply(&thereply, &len)) 
				 || rp_isbad(thereply.rp_val)) {
		mm_end(NOTOK);
		ml_state = ML_FRESH;
		return(NOTOK);
	    }
	    ml_state = ML_MSG;
	    break;

	case NOTOK:
	    mm_end(NOTOK);
	    ml_state = ML_FRESH;
	    break;
    }
    return(OK);
}
/**/

int ml_1adr (ret, sndr, from, sub, adr)
				  /* all set-up overhead in 1 proc      */
int     ret,                      /* allow return mail to sender?       */
	sndr;                     /* include Sender field?              */
char    *sub,                     /* subject line                       */
	*from,                    /* from field                         */
	*adr;                     /* the one address to receive msg     */
{

    if (ml_init (ret, sndr, from, sub) != OK ||
	    ml_adr (adr) != OK ||
	    ml_aend () != OK ||
	    ml_tinit () != OK)
	return (NOTOK);

    return (OK);
}
